planner/cascades/transformation_rules.go

```diff
@@ -1281,6 +1282,55 @@ func (r *PushTopNDownTiKVSingleGather) OnTransform(old *memo.ExprIter) (newExprs
 	return []*memo.GroupExpr{finalTopNExpr}, true, false, nil
 }
 
+// MergeAdjacentTopN merge adjacent TopN.
+type MergeAdjacentTopN struct {
+	baseRule
+}
+
+// NewRuleMergeAdjacentTopN creates a new Transformation MergeAdjacentTopN.
+// The pattern of this rule is `TopN->TopN->X`.
+func NewRuleMergeAdjacentTopN() Transformation {
+	rule := &MergeAdjacentTopN{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandTopN,
+		memo.EngineAll,
+		memo.NewPattern(memo.OperandTopN, memo.EngineAll),
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+func (r *MergeAdjacentTopN) Match(expr *memo.ExprIter) bool {
+	topN := expr.GetExpr().ExprNode.(*plannercore.LogicalTopN)
+	child := expr.Children[0].GetExpr().ExprNode.(*plannercore.LogicalTopN)
+	if len(topN.ByItems) != len(child.ByItems) {
+		return false
+	}
+	for i := 0; i < len(topN.ByItems); i++ {
+		if (*topN.ByItems[i]) != (*child.ByItems[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+// OnTransform implements Transformation interface.
+// This rule tries to merge adjacent TopN.
+func (r *MergeAdjacentTopN) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	topN := old.GetExpr().ExprNode.(*plannercore.LogicalTopN)
+	child := old.Children[0].GetExpr().ExprNode.(*plannercore.LogicalTopN)
+	childGroups := old.Children[0].GetExpr().Children
+
+	newTopN := plannercore.LogicalTopN{
+		Count:   topN.Count,
+		Offset:  topN.Offset + child.Offset,
+		ByItems: child.ByItems,
+	}.Init(child.SCtx(), child.SelectBlockOffset())
```

Comment:
I think this is still not correct. For example, `select * from (select * from t order by a limit 3) t1 order by a limit 5)`. The child `TopN` only returns 3 rows, so the top one can get at most 3 rows.

Reply:
Also, if `topN.Offset >= child.Count`, we can simplify the `TopN` to a `TableDual`.

Reply:
#14306 might be similar to this rule.

Reply:
@francis0407 Update. PTAL